### PR TITLE
[SPARK-42040][SQL] SPJ: Introduce a new API for V2 input partition to report partition statistics

### PR DESCRIPTION
**Pull request by zhuqi-lucas**
### What changes were proposed in this pull request?
This PR introduces a new V2 min-in HasPartitionStatistics which can be used to return the partition statistics of a InputPartition.

### Why are the changes needed?
As part of the Storage Partitioned Join work ([SPIP](https://issues.apache.org/jira/browse/SPARK-37166)), we'll need to introduce a way for a V2 InputPartition to return its partition size,  It's useful for a InputPartition to also report its size (in bytes), so that Spark can use the info to decide whether partition grouping should be applied or not. 

This will be used later in follow-up PRs.

### Does this PR introduce _any_ user-facing change?
Yes, a new V2 mix-in HasPartitionStatistics will be introduced.

### How was this patch tested?
Extended InMemoryTable to support this new interface, and added a new unit test to verify the API, and the mock the partition statistics to test.

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

**Comment by zhuqi-lucas** [comment]
cc @sunchao I try to help this task, Could you take a look? Thanks!

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionSize.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * A mix-in for input partitions whose records are clustered on the same set of partition keys
+ * (provided via {@link SupportsReportPartitioning}, see below). Data sources can opt-in to
+ * implement this interface for the partitions they report to Spark, which will use the info
+ * to decide whether partition grouping should be applied or not.
+ *
+ * @see org.apache.spark.sql.connector.read.SupportsReportPartitioning
+ * @since 4.0.0
+ */
+public interface HasPartitionSize extends InputPartition {
```
Comment (sunchao) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionSize.java:31]: I wonder if we can make this more general and support partition level stats as well, like number of rows.

**Review by sunchao — COMMENTED** [review]
cc @aokolnychyi @RussellSpitzer @rdblue do you think this could be useful for Iceberg to pass partition stats to Spark? SPJ could leverage this to make better decisions on how to combine partitions (like which side to choose during partially clustered distribution), but I'm not sure whether there are more use cases.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionSize.java (hunk 1 above)
Comment (zhuqi-lucas) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionSize.java:31]: Thank you @sunchao for review, and this is a good suggestion, i check the Iceberg code, it includes the sizeBytes and estimatedRowsCount, filesCount.  Let me address this!

```
  @Override
  default long sizeBytes() {
    return tasks().stream().mapToLong(ScanTask::sizeBytes).sum();
  }

  @Override
  default long estimatedRowsCount() {
    return tasks().stream().mapToLong(ScanTask::estimatedRowsCount).sum();
  }

  @Override
  default int filesCount() {
    return tasks().stream().mapToInt(ScanTask::filesCount).sum();
  }
}
```

Comment (zhuqi-lucas) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionSize.java]: Addressed in latest PR.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * to decide whether partition grouping should be applied or not.
+ *
+ * @see org.apache.spark.sql.connector.read.SupportsReportPartitioning
+ * @since 4.0.0
+ */
+public interface HasPartitionStatistics extends InputPartition {
+  /**
+   * Returns the value of the partition statistics associated to this partition.
+   */
+  OptionalLong partitionSizeInBytes();
```
Comment (sunchao) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java:35]: nit: can we just remove `partition` and just use `sizeInBytes`? it's pretty obvious this is for partition statistics.

**Review by sunchao — APPROVED** [review]
LGTM with one nit, cc @cloud-fan too

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java:35]: +1

Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java:35]: ditto for other methods.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  OptionalLong partitionNumRows();
```
Comment (beliefer) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java:36]: As @sunchao said, we can simplify the name for other methods.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java (hunk 2 above)
Comment (zhuqi-lucas) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java]: Thank you @sunchao @cloud-fan , good suggestion, addressed in latest PR.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java (hunk 3 above)
Comment (zhuqi-lucas) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java]: Thank you @beliefer , addressed this in latest PR.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * @see org.apache.spark.sql.connector.read.SupportsReportPartitioning
+ * @since 4.0.0
+ */
+public interface HasPartitionStatistics extends InputPartition {
+  /**
+   * Returns the value of the partition statistics associated to this partition.
+   */
+  OptionalLong sizeInBytes();
+  OptionalLong numRows();
```
Comment (sunchao) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java:36]: @zhuqi-lucas could we add some comments for `numRows` and `fileCount` too?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ * (provided via {@link SupportsReportPartitioning}, see below). Data sources can opt-in to
+ * implement this interface for the partitions they report to Spark, which will use the info
+ * to decide whether partition grouping should be applied or not.
+ *
+ * @see org.apache.spark.sql.connector.read.SupportsReportPartitioning
+ * @since 4.0.0
+ */
+public interface HasPartitionStatistics extends InputPartition {
+  /**
+   * Returns the value of the partition statistics associated to this partition.
```
Comment (sunchao) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java:33]: Hmm I think this comment is not correct?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java (hunk 4 above)
Comment (zhuqi-lucas) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java:39]: Thank you @sunchao for this suggestion, addressed latest PR.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java (hunk 5 above)
Comment (zhuqi-lucas) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java]: Addressed latest PR.

**Comment by szehon-ho** [comment]
> cc @aokolnychyi @RussellSpitzer @rdblue do you think this could be useful for Iceberg to pass partition stats to Spark? SPJ could leverage this to make better decisions on how to combine partitions (like which side to choose during partially clustered distribution), but I'm not sure whether there are more use cases.

@sunchao   Aside from picking the side of partially clustered distribution, would we also be able to use it to group smaller partitions?  Example a table is partition by date, and older days have not much data (on both sides), group many of the older days into the same task.

Similar to AQE coalesce partitions, but it looks like it applies only after shuffle, so looks like it doesnt apply for SPJ?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * @see org.apache.spark.sql.connector.read.SupportsReportPartitioning
+ * @since 4.0.0
+ */
+public interface HasPartitionStatistics extends InputPartition {
+  /**
+   * Returns the size in bytes of the partition statistics associated to this partition.
+   */
+  OptionalLong sizeInBytes();
+  /**
```
Comment (szehon-ho) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java:36]: Nit: can we add some newline between method and next line?

Comment (zhuqi-lucas) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/HasPartitionStatistics.java:38]: Thank you @szehon-ho for review, addressed in latest PR.

**Comment by sunchao** [comment]
> @sunchao Aside from picking the side of partially clustered distribution, would we also be able to use it to group smaller partitions? Example a table is partition by date, and older days have not much data (on both sides), group many of the older days into the same task.

Yea I I think that would be an interesting use case. If we know the partitions from both sides of the join AND the size for each partition, we can probably make some better decisions. 

> Similar to AQE coalesce partitions, but it looks like that applies only after shuffle, so looks like it doesnt apply for SPJ?

Right, this doesn't to SPJ.

**Comment by sunchao** [comment]
Thanks, merged to master!

**Pull request closed without merging**